pyanaconda/bootloader/grub2.py

```diff
@@ -114,7 +114,7 @@ class GRUB2(BootLoader):
     stage2_must_be_primary = False
 
     # requirements for boot devices
-    stage2_device_types = ["partition", "mdarray"]
+    stage2_device_types = ["partition", "mdarray", "btrfs subvolume"]
     stage2_raid_levels = [raid.RAID0, raid.RAID1, raid.RAID4,
                           raid.RAID5, raid.RAID6, raid.RAID10]
     stage2_raid_member_types = ["partition"]
```

Comment:
This statement confused me. This didn't exist before, and I'm not entirely sure what it does. Should I be concerned about it for Btrfs enablement?

Reply:
stage2 device is a reference to the portion of BIOS GRUB found in /boot/grub (/boot/grub2 on Fedora). There are three non-fatal consequences with /boot on Btrfs:
a. If the user opts for zstd compression, boot menu will fail to populate as GRUB+blscfg.mod will find but fail to read the compressed config files in /boot/loader/entries/*.conf. Fix for that is to get zstd baked into the grubx64.efi on UEFI and stage 1.5 for BIOS GRUB. I have a todo for that. (lzs and zlib compression support is built into btrfs.mod, so that's always worked). The short term work around is for the user to use `chattr +C` on /boot
b. On BIOS, the grubenv is located on /boot/grub2/grubenv and GRUB disallows writes to grubenv when on btrfs (and raid, lvm, luks, zfs, maybe some other things) because GRUB's file writes are sector overwrites, there's no file system metadata updates, including checksums on Btrfs. Mismatching checksums means kernel code will EIO on such files. Basically Btrfs considers GRUB writes as corrupting file data. And it's something other fs upstreams take issue with including XFS and ext4. So a better solution for getting state out of GRUB is needed. The problem is the hidden GRUB menu feature includes GRUB changing boot_success=1 to 0, and then only if boot success is grubenv modified to change boot_success=1. That way GRUB knows whether to display the GRUB menu following a failed boot. On Btrfs, the menu won't ever be shown unless the user intervenes by pressing a key, because GRUB won't write to files on Btrfs.
c. There's this small complication with the BootloaderSpec which ostensibly is about having a shared /boot across multiple file systems; and I'm not sure yet if this is something Fedora wants to push forward on, and if so I'm not really sure how to support it on Btrfs because we don't have subvolume type UUIDs like we do for GPT partitions. We could maybe use an xattr but that's a bit "secret decoder ring". See bug 
https://bugzilla.redhat.com/show_bug.cgi?id=825236#c38

Reply:
> Fix for that is to get zstd baked into the grubx64.efi on UEFI

This one should be easy to fix then.